cdap-app-fabric/src/main/java/co/cask/cdap/internal/provision/DeprovisionTask.java

```diff
@@ -33,78 +38,173 @@
+                                     taskInfo.getCluster());
+          persistClusterInfo(deletingInfo);
+          execute(deletingInfo);
+          break;
+        case CREATING:
+        case FAILED:
+        case ORPHANED:
+          LOG.warn("Error deprovisioning cluster for program run {}. Cluster will be moved to orphaned state.",
+                   programRunId);
+          provisionerNotifier.orphaned(programRunId);
+          ProvisioningTaskInfo failedInfo =
+            new ProvisioningTaskInfo(taskInfo, new ClusterOp(ClusterOp.Type.DEPROVISION, ClusterOp.Status.ORPHANED),
+                                     taskInfo.getCluster());
+          persistClusterInfo(failedInfo);
           break;
       }
 
     } catch (Throwable t) {
-      LOG.warn("Error deprovisioning cluster for program run {}", programRunId, t);
-      // TODO: CDAP-13246 handle retries
+      Throwables.propagateIfInstanceOf(t, InterruptedException.class);
+      handleFailure(taskInfo, t);
+    }
+  }
+
+  private void deleteCluster(ProvisioningTaskInfo taskInfo) throws InterruptedException {
+    try {
+      Cluster cluster = taskInfo.getCluster();
+      Retries.callWithRetries(() -> {
+        provisioner.deleteCluster(provisionerContext, cluster);
+        return null;
+        }, retryStrategy, e -> e instanceof RetryableProvisionException);
+      ClusterOp op = new ClusterOp(ClusterOp.Type.DEPROVISION, ClusterOp.Status.POLLING_DELETE);
+      Cluster updatedCluster = new Cluster(cluster == null ? null : cluster.getName(), ClusterStatus.DELETING,
+                                           cluster == null ? Collections.emptyList() : cluster.getNodes(),
+                                           cluster == null ? Collections.emptyMap() : cluster.getProperties());
+      ProvisioningTaskInfo pollingInfo = new ProvisioningTaskInfo(taskInfo, op, updatedCluster);
+      persistClusterInfo(pollingInfo);
+      execute(pollingInfo);
+    } catch (Throwable t) {
+      Throwables.propagateIfInstanceOf(t, InterruptedException.class);
+      handleFailure(taskInfo, t);
+    }
+  }
+
+  private void handleFailure(ProvisioningTaskInfo taskInfo, Throwable t) {
+    LOG.warn("Error deprovisioning cluster for program run {} during {} step. Cluster will be moved to orphaned state.",
+             programRunId, taskInfo.getClusterOp().getStatus(), t);
+    provisionerNotifier.orphaned(programRunId);
+    ProvisioningTaskInfo failedInfo =
+      new ProvisioningTaskInfo(taskInfo, new ClusterOp(ClusterOp.Type.DEPROVISION, ClusterOp.Status.FAILED),
+                               taskInfo.getCluster());
+    persistClusterInfo(failedInfo);
+  }
+
+  // write the state of the task to the ProvisionerDataset, retrying if any exception is caught
+  private void persistClusterInfo(ProvisioningTaskInfo taskInfo) {
+    try {
+      // retry on every exception, up to the retry limit
```

Comment:
I think we shouldn't retry on InterruptedException of if the thread is interrupted?

Reply:
right, will change